bugfix: CLDSRV-361 fix exception with batch delete of null version

jonathan-gramain opened the pull request:
Fix an exception in Cloudserver when doing a batch delete containing a deletion of a null version that does not exist on a versioned object

jonathan-gramain committed:
bugfix: CLDSRV-361 fix exception with batch delete of null version

- Fix an exception in Cloudserver when doing a batch delete containing a
  deletion of a null version that does not exist on a versioned object

- The changes also fix the return code when fetching a noncurrent null
  version that was deleted: in such case it was returning NoSuchKey,
  now returns NoSuchVersion per spec.

The change consists of:

- consolidate and fix the API contract of metadataGetObject() and
  metadataValidateBucketAndObj() for "not found" objects or object
  versions: there was a mix of return codes from Metadata (NoSuchKey)
  and an OK response with no metadata returned in others, depending on
  the exact scenario and object state. Fixed by always returning an OK
  response and no metadata if the target version is not found, to let
  the caller set the appropriate error code in the API response

- for multiObjectDelete, fix by handling the OK response with empty
  object metadata response specifically as a "not found" case, like
  for other API calls

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-361 contains:


* `7.10.27`

* `7.70.19`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.27`

* `7.70.19`

* `8.4.20`

* `8.5.21`

* `8.6.9`

* `8.7.10`


Please check the `Fix Version/s` of CLDSRV-361, or the target
branch of this pull request.

jonathan-gramain commented:
ping

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion` into
`development/7.10`
* [w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/7.70...w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/7.70` 
* [w/8.4/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/8.4...w/8.4/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/8.4` 
* [w/8.5/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/8.5` 
* [w/8.6/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/8.6` 
* [w/8.7/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# History mismatch



Merge commit #e65fe913a01fa26befa3758ca768d12644cdb2e7 on the integration branch
`w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion` is merging a branch which is neither the current
branch `bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion` nor the development branch
`development/7.70`.

It is likely due to a rebase of the branch `bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain committed:
CLDSRV-361 [8.x fix] backbeat multiple backend route fix

Fix handling of response of metadataGetObject() in multiple backend
PutTagging and DeleteTagging when target object does not exist: return
error code NoSuchKey in this case.

Note: NoSuchKey used to be returned by metadataGetObject() helper but
not always, now it's never returned, instead the "objMD" returned
value is null.

jonathan-gramain commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion` with contents from `bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion`
and `development/7.70`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion origin/development/7.70
 $ git merge origin/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion` with contents from `w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion origin/development/8.4
 $ git merge origin/w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion` into
`development/7.10`
* [w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/7.70...w/7.70/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/7.70` 
* [w/8.4/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/8.4...w/8.4/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/8.4` 
* [w/8.5/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/8.5` 
* [w/8.6/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/8.6` 
* [w/8.7/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/bugfix/CLDSRV-361-fixExceptionBatchDeleteNullVersion) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

tmacro commented on lib/metadata/metadataUtils.js:
```diff
@@ -26,7 +26,13 @@ function getNullVersion(objMD, bucketName, objectKey, log, cb) {
         options.versionId = objMD.nullVersionId;
-        return metadata.getObjectMD(bucketName, objectKey, options, log, cb);
+        return metadata.getObjectMD(bucketName, objectKey, options, log, (err, nullVersionMD) => {
+            if (err && err.is.NoSuchKey) {
```
Not sure it its possible for it to be a non-Arsenal error, but better safe than sorry.
```suggestion
            if (err && err.is?.NoSuchKey) {
```

tmacro commented on lib/metadata/metadataUtils.js:
```diff
@@ -96,6 +102,10 @@ function metadataGetObject(bucketName, objectKey, versionId, log, cb) {
     };
     return metadata.getObjectMD(bucketName, objectKey, options, log,
         (err, objMD) => {
+            if (err && err.is.NoSuchKey) {
```
```suggestion
            if (err && err.is?.NoSuchKey) {
```

KazToozs reviewed (approved):
LGTM

jonathan-gramain committed:
CLDSRV-361 guard before accessing err.is field

jonathan-gramain commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.4`



Please check the status of the associated issue CLDSRV-361.

Goodbye jonathan-gramain.
